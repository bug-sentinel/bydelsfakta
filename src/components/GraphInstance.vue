<template>
  <div class="graph__shadow" role="tabpanel" :class="shadowClass">
    <div
      class="graph__container"
      :class="{ hidden: mode === 'table' }"
      ref="graphContainer"
      aria-hidden="true"
      @scroll="drawShadows"
      v-dragscroll
    >
      <div class="spinner" v-if="loading">
        <img src="../assets/spinner.svg" alt="" />
        <span class="spinner__text">{{ $t('loading.loadingData') }}</span>
      </div>
      <div class="error" v-if="error">
        <h2>{{ errorMessage }}</h2>
      </div>
      <svg
        v-if="!error"
        class="graph__svg"
        aria-hidden="true"
        ref="svg"
        :class="{ loading }"
        @click="showHelp = false"
      ></svg>
    </div>
    <div
      :class="{ 'visually-hidden': mode === 'graph' }"
      class="graph__tablecontainer"
      ref="tableContainer"
      v-dragscroll
      @scroll="drawShadows"
    >
      <table>
        <caption></caption>
        <thead></thead>
        <tbody></tbody>
      </table>
    </div>
    <!-- <table class="visually-hidden"> -->
    <resize-observer @notify="handleResize"></resize-observer>
    <div class="help" v-if="settings.help && mode === 'graph'">
      <button class="help__button" @click="showHelp = !showHelp">
        <i aria-hidden="true" class="material-icons">help_outline</i>
      </button>
      <div v-if="showHelp" class="help__text">{{ settings.help }}</div>
    </div>
  </div>
</template>

<script>
import * as d3 from 'd3';
import { mapState } from 'vuex';
import { dragscroll } from 'vue-dragscroll';
import TemplateA from '../util/graph-templates/templateA';
import TemplateB from '../util/graph-templates/templateB';
import TemplateC from '../util/graph-templates/templateC';
import TemplateD from '../util/graph-templates/templateD';
import TemplateE from '../util/graph-templates/templateE';
import TemplateF from '../util/graph-templates/templateF';
import TemplateG from '../util/graph-templates/templateG';
import TemplateH from '../util/graph-templates/templateH';
import TemplateI from '../util/graph-templates/templateI';
import TemplateJ from '../util/graph-templates/templateJ';
import districtNames from '../config/districtNames';

export default {
  data: () => ({
    svg: false,
    shadow: {
      left: false,
      right: false,
    },
    data: null,
    loading: true,
    currentTemplate: false,
    allDistricts: false,
    error: false,
    errorMessage: '',
    showHelp: false,
  }),

  computed: {
    shadowClass() {
      let str = '';
      if (this.shadow.left && this.mode === 'graph') str += ' graph__shadow--left';
      if (this.shadow.right) str += ' graph__shadow--right';
      return str;
    },
    ...mapState(['districts', 'compareDistricts']),
    filteredData() {
      if (!this.compareDistricts || this.districts.includes('alle')) return this.data;

      const selectedDistrictNames = this.districts.map(id => districtNames[id]);

      return {
        meta: this.data.meta,
        data: this.data.data.filter(d => {
          return selectedDistrictNames.includes(d.geography) || d.totalRow;
        }),
      };
    },
  },

  watch: {
    settings: function() {
      this.draw();
    },
    districts: function(to, from) {
      if (from.length > 1 && (to.length > 1 || this.compareDistricts)) {
        this.draw({ keepData: true });
      } else {
        this.draw();
      }
    },
  },

  mounted() {
    this.draw();
  },

  methods: {
    drawShadows() {
      let ref;
      if (this.mode === 'graph') {
        ref = 'graphContainer';
      } else if (this.mode === 'table') {
        ref = 'tableContainer';
      } else {
        return;
      }

      const width = this.$refs[ref].parentElement.clientWidth;
      const spaceLeft = this.$refs[ref].scrollLeft;
      const spaceRight = this.$refs[ref].scrollWidth - width - spaceLeft;

      this.shadow.left = spaceLeft > 5;
      this.shadow.right = spaceRight > 5;
    },

    handleResize() {
      if (this.loading) return;
      this.svg.resize(this.data, { method: this.settings.method });
      this.drawShadows();
    },

    async draw(options = {}) {
      this.loading = true;

      const geoParam = this.compareDistricts ? '00' : this.districts[0];
      if (!options.keepData) {
        this.data = await d3
          .json(`${this.settings.url}?geography=${geoParam}`)
          .then(data => {
            data.data.map(district => {
              district.geography = districtNames[district.geography];
              return district;
            });
            return data;
          })
          .catch(() => {
            this.error = true;
            this.errorMessage = this.$t('error.connectionLost');
            this.loading = false;
          });
      }

      if (this.data) {
        this.error = false;
        this.errorMessage = '';
      } else {
        return;
      }

      if (this.currentTemplate !== this.settings.template && !options.keepData) {
        switch (this.settings.template) {
          case 'a':
            this.svg = new TemplateA(this.$refs['svg']);
            break;
          case 'b':
            this.svg = new TemplateB(this.$refs['svg']);
            break;
          case 'c':
            this.svg = new TemplateC(this.$refs['svg']);
            break;
          case 'd':
            this.svg = new TemplateD(this.$refs['svg']);
            break;
          case 'e':
            this.svg = new TemplateE(this.$refs['svg']);
            break;
          case 'f':
            this.svg = new TemplateF(this.$refs['svg']);
            break;
          case 'g':
            this.svg = new TemplateG(this.$refs['svg']);
            break;
          case 'h':
            this.svg = new TemplateH(this.$refs['svg']);
            break;
          case 'i':
            this.svg = new TemplateI(this.$refs['svg']);
            break;
          case 'j':
            this.svg = new TemplateJ(this.$refs['svg']);
            break;
          default:
            break;
        }
      }
      this.currentTemplate = this.settings.template;

      this.loading = false;

      this.svg.render(this.filteredData, {
        method: this.settings.method,
        initialRender: true,
        compareDistricts: this.compareDistricts,
      });
      this.drawShadows();
    },
  },
  props: {
    settings: {
      type: Object,
      required: true,
    },
    mode: {
      type: String,
      required: true,
    },
  },
  directives: {
    dragscroll,
  },
};
</script>

<style lang="scss" scoped>
@import './../styles/colors';
@import './../styles/variables';

.graph {
  &__shadow {
    position: relative;

    &::before,
    &::after {
      bottom: 0;
      content: '';
      display: block;
      opacity: 0;
      pointer-events: none;
      position: absolute;
      top: 0;
      transition: opacity 0.1s ease-in-out;
      width: 2em;
      z-index: 1;
    }

    &::before {
      background: linear-gradient(to right, rgba(black, 0.3), rgba(black, 0.1) 60%, rgba(black, 0) 100%);
      left: 0;
    }

    &::after {
      background: linear-gradient(to left, rgba(black, 0.3), rgba(black, 0.1) 60%, rgba(black, 0) 100%);
      right: 0;
    }

    &--left::before {
      opacity: 1;
    }

    &--right::after {
      opacity: 1;
    }
  }

  &__container {
    overflow: auto;
    padding: 1em 1em 0 1em;
    position: relative;
  }

  &__tablecontainer {
    overflow-x: auto;
    position: relative;
    width: 100%;
  }

  &__svg {
    &.loading {
      animation-delay: 0.5s;
      animation-duration: 1s;
      animation-fill-mode: forwards;
      animation-iteration-count: 1;
      animation-name: fadeOut;
      pointer-events: none;
    }
  }
}

.spinner {
  align-items: center;
  animation-delay: 0.5s;
  animation-duration: 1s;
  animation-fill-mode: forwards;
  animation-iteration-count: 1;
  animation-name: fadeIn;
  background-color: rgba(black, 0.05);
  display: flex;
  flex-direction: column;
  height: 100%;
  justify-content: center;
  left: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  width: 100%;
  z-index: 999;

  &__text {
    padding: 0.75em;
  }
}

.tick text {
  color: $color-grey-600;
  font-family: monospace;
  font-size: $font-small;
  font-weight: 500;
  text-rendering: geometricPrecision;
}

.help {
  position: absolute;
  top: 0.5em;
  right: 0.5em;
  z-index: 1;

  &__button {
    right: 0;
    top: 0;
    height: 3em;
    width: 3em;
    display: flex;
    border-radius: 50%;
    position: absolute;
    color: $color-purple;
    cursor: pointer;
    align-items: center;
    justify-content: center;

    & > i {
      pointer-events: none;
      font-size: 1.5em;
    }

    &:hover {
      background: $color-blue;
    }

    &::before {
      content: '';
      display: block;
      position: absolute;
      top: -0.5rem;
      right: -0.5rem;
      height: 4rem;
      width: 4rem;
      background: white;
      z-index: -1;
    }
  }

  &__text {
    padding: 1em;
    background: $color-purple;
    color: $color-yellow;
    font-weight: 500;
    position: absolute;
    top: 3em;
    right: 0;
    border-radius: 3px;
    width: 250px;
    box-shadow: 0 3px 4px rgba($color-purple, 0.5);

    &::before {
      border-bottom: 0.75em solid $color-purple;
      border-left: 0.75em solid transparent;
      border-right: 0.75em solid transparent;
      top: -0.5em;
      content: '';
      display: block;
      position: absolute;
      right: 0.75em;
    }
  }
}

@keyframes fadeIn {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

@keyframes fadeOut {
  from {
    opacity: 1;
  }

  to {
    opacity: 0;
  }
}
</style>
