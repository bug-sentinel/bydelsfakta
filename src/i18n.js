import Vue from 'vue';
import VueI18n from 'vue-i18n';

function loadLocaleMessages() {
  const locales = require('./locales/nb-no.json');
  return {
    'nb-no': locales,
  };
}

export default function setupI18n() {
  Vue.use(VueI18n);

  return new VueI18n({
    locale: process.env.VUE_APP_I18N_LOCALE || 'nb-no',
    fallbackLocale: process.env.VUE_APP_I18N_FALLBACK_LOCALE || 'nb-no',
    messages: loadLocaleMessages(),
  });
}
