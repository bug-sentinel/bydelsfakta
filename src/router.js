import { createRouter, createWebHistory } from 'vue-router';

import { topicNames, disabledTopics } from './config/topics';
import allDistricts from './config/allDistricts';

export const routes = [
  {
    path: '/',
    name: 'Home',
    component: () => import('@/views/District.vue'),
  },
  {
    path: '/bydel/:district?',
    name: 'District',
    component: () => import('@/views/District.vue'),
    props: true,
  },
  {
    path: '/bydel/:district/:topic',
    name: 'Topic',
    component: () => import('@/views/Topic.vue'),
    props: true,
  },
  {
    path: '/:pathMatch(.*)*',
    name: 'NotFound',
    component: () => import('@/views/NotFound.vue'),
  },
];

const router = createRouter({
  history: createWebHistory(),
  routes,
});

// Temp while we find out what our homepage should look like
router.beforeEach((to, from) => {
  if (to.params.topic) {
    if (
      (!topicNames.find((name) => name === to.params.topic) && import.meta.env.VITE_PRODUCTION_DATA === 'prod') ||
      disabledTopics.includes(to.params.topic)
    ) {
      return { name: 'NotFound', params: [to.path] };
    }
  }
  if (to.params.district) {
    const districts = to.params.district.split('-');
    if (districts[0] === 'alle') {
      return true;
    }
    if (districts.length === 1) {
      if (allDistricts.find((district) => district.uri === districts[0])) {
        return true;
      }
      if (allDistricts.find((district) => district.key === districts[0])) {
        return true;
      }
      return { name: 'NotFound', params: [to.path] };
    }
    if (districts.length > 1) {
      const errors = [];
      districts.forEach((district) => {
        if (!allDistricts.find((item) => item.key === district)) {
          errors.push(district);
        }
      });
      if (errors.length > 0) {
        return { name: 'NotFound', params: [to.path] };
      }
    }
  }
  if (to.params.topic !== from.params.topic || (to.name === 'Topic' && from.name === 'District')) {
    window.scroll(0, 0);
  }
  return true;
});

export default router;
