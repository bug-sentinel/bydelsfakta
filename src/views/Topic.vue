<template>
  <div class="main-container">
    <div class="main-container__cards">
      <template v-for="(card, cardIndex) in topics[topic].cards">
        <graph-card
          :key="`card-${cardIndex}`"
          :settings="card"
          v-if="
            card.show === productionMode && card.show !== null
              ? true
              : card.show === true && productionMode === false
              ? true
              : productionMode === null
          "
        />
      </template>
      <div class="related">
        <h2 class="section-heading">{{ $t('topic.seeMore') }}</h2>
        <div class="topics-grid">
          <v-category
            v-for="(item, relatedIndex) in topics[`${topic}`].related"
            :key="`related-${relatedIndex}`"
            :id="topics[`${item}`].value"
            :category="topics[`${item}`].options.kategori"
            :topic="topics[`${item}`].options.tema"
            :bg-image="topics[`${item}`].options.bgImage"
            :txt-color="topics[`${item}`].options.txtColor"
            :district="$route.params.district"
            :disabled="
              topics[`${topic}`].show === productionMode && topics[`${topic}`].show !== null
                ? false
                : topics[`${topic}`].show === true && productionMode === false
                ? false
                : productionMode !== null
            "
          />
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { mapGetters, mapState } from 'vuex';
import VCategory from '../components/VCategory.vue';
import GraphCard from '../components/GraphCard.vue';

import { topics, disabledTopics } from '../config/topics';

export default {
  name: 'Topic',
  components: {
    GraphCard,
    VCategory,
  },
  props: {
    district: {
      type: String,
      required: true,
    },
    topic: {
      type: String,
      required: true,
    },
  },

  data() {
    return {
      topics: topics,
      disabledTopics: disabledTopics,
    };
  },

  computed: {
    ...mapState(['productionMode']),
    ...mapGetters(['geoDistricts']),
  },
};
</script>

<style scoped>
.related {
  width: 100%;
}
</style>
