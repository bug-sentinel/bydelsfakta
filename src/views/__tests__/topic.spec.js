import { mount } from '@vue/test-utils';
import { createStore } from 'vuex';
import { createRouter, createWebHistory } from 'vue-router';
import Vue3Resize from 'vue3-resize';
import clickOutside from '@/directives/clickOutside';
import i18n from '@/i18n';
import { routes } from '@/router';
import mockStore from '@/../tests/MockStore';
import GraphInstance from '@/components/GraphInstance.vue';
import Topic from '../Topic.vue';

describe('Topic', () => {
  let wrapper = null;
  let router = null;
  let store = null;

  beforeEach(async () => {
    store = createStore(mockStore);
    router = createRouter({
      history: createWebHistory(),
      routes,
    });

    GraphInstance.methods.draw = jest.fn();

    wrapper = mount(Topic, {
      global: {
        plugins: [router, store, i18n, Vue3Resize],
        directives: {
          'click-outside': clickOutside,
        },
        stubs: {
          'ok-icon': true,
          spinner: true,
          'v-category': true,
        },
      },
      props: {
        district: 'gamleoslo',
        topic: 'alder',
      },
    });
    await router.push('/bydel/gamleoslo/alder');
  });

  afterEach(() => {
    wrapper.unmount();
  });

  test('renders topic-component and finds main-container class', () => {
    expect(wrapper.classes('main-container')).toBe(true);
  });

  test('renders component correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
