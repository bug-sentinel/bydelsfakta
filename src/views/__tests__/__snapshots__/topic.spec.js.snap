// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Topic renders component correctly 1`] = `
<div
  class="main-container"
>
  <div
    class="main-container__cards"
  >
    
    <section
      class="card-container large"
    >
      <div
        class="card"
      >
        <!--v-if-->
        <header
          class="card__header"
        >
          <div
            class="card__headertext"
          >
            <h2
              class="card__title"
            >
              Befolkningen etter alder
            </h2>
          </div>
          <nav
            class="card__nav"
          >
            <div
              class="tabs"
              role="tablist"
            >
              <div
                class="vue3-resize-observer"
                tabindex="-1"
              >
                <object
                  aria-hidden="true"
                  data="about:blank"
                  tabindex="-1"
                  type="text/html"
                />
              </div>
              
              <button
                aria-label="Antall"
                class="active tabs__tab--hidden tabs__tab"
                data-id="tabButton-Befolkningen etter alder-0"
                role="tab"
              >
                Antall
              </button>
              <button
                aria-label="Andel"
                class="tabs__tab--hidden tabs__tab"
                data-id="tabButton-Befolkningen etter alder-1"
                role="tab"
              >
                Andel
              </button>
              
              <select
                aria-hidden=""
                class="select tabs__select"
              >
                
                <option
                  value="0"
                >
                  Antall
                </option>
                <option
                  value="1"
                >
                  Andel
                </option>
                
              </select>
            </div>
            <div
              class="card__toggle-menu"
            >
              <button
                aria-label="Graf"
                class="card__toggle-button card__toggle-button--active"
                title="Graf"
              >
                <ok-icon-stub
                  class="context-menu__dropdown-item-icon"
                  iconref="graph"
                  options="[object Object]"
                />
              </button>
              <button
                aria-label="Tabell"
                class="card__toggle-button"
                title="Tabell"
              >
                <ok-icon-stub
                  class="context-menu__dropdown-item-icon"
                  iconref="table"
                  options="[object Object]"
                />
              </button>
              <!--v-if-->
            </div>
            <div
              class="context-menu"
            >
              <button
                aria-haspopup="true"
                aria-label="Åpne meny"
                class="context-menu__button"
                data-id="context-menu-button"
                title="Åpne meny"
              >
                <ok-icon-stub
                  iconref="menu"
                  options="[object Object]"
                />
                <span
                  class="button-label"
                >
                  Valg
                </span>
              </button>
              <!--v-if-->
            </div>
          </nav>
        </header>
        <div
          class="graph__shadow"
          role="tabpanel"
        >
          <div
            aria-hidden="true"
            class="graph__container"
          >
            <div
              class="spinner"
            >
              <spinner-stub />
              <span
                class="spinner__text"
              >
                Henter data
              </span>
            </div>
            <!--v-if-->
            <svg
              aria-hidden="true"
              class="graph__svg loading"
            />
          </div>
          <div
            class="visually-hidden graph__tablecontainer"
          >
            <h3
              class="table-heading"
            />
            <table
              class=""
            >
              <thead />
              <tbody />
            </table>
            <!--v-if-->
          </div>
          <div
            class="vue3-resize-observer"
            tabindex="-1"
          >
            <object
              aria-hidden="true"
              data="about:blank"
              tabindex="-1"
              type="text/html"
            />
          </div>
          <div
            class="help"
          >
            <button
              aria-label="Vis hjelpetekst"
              class="help__button"
            >
              <ok-icon-stub
                iconref="question"
                options="[object Object]"
              />
            </button>
            <!--v-if-->
          </div>
        </div>
        <!--v-if-->
        <!--v-if-->
      </div>
    </section>
    <section
      class="card-container large"
    >
      <div
        class="card"
      >
        <!--v-if-->
        <header
          class="card__header"
        >
          <div
            class="card__headertext"
          >
            <h2
              class="card__title"
            >
              Befolkningspyramide
            </h2>
          </div>
          <nav
            class="card__nav"
          >
            <div
              class="tabs"
              role="tablist"
            >
              <div
                class="vue3-resize-observer"
                tabindex="-1"
              >
                <object
                  aria-hidden="true"
                  data="about:blank"
                  tabindex="-1"
                  type="text/html"
                />
              </div>
              
              <button
                aria-label="Status"
                class="active tabs__tab--hidden tabs__tab"
                data-id="tabButton-Befolkningspyramide-0"
                role="tab"
              >
                Status
              </button>
              
              <select
                aria-hidden=""
                class="select tabs__select"
              >
                
                <option
                  value="0"
                >
                  Status
                </option>
                
              </select>
            </div>
            <div
              class="card__toggle-menu"
            >
              <button
                aria-label="Graf"
                class="card__toggle-button card__toggle-button--active"
                title="Graf"
              >
                <ok-icon-stub
                  class="context-menu__dropdown-item-icon"
                  iconref="graph"
                  options="[object Object]"
                />
              </button>
              <button
                aria-label="Tabell"
                class="card__toggle-button"
                title="Tabell"
              >
                <ok-icon-stub
                  class="context-menu__dropdown-item-icon"
                  iconref="table"
                  options="[object Object]"
                />
              </button>
              <!--v-if-->
            </div>
            <div
              class="context-menu"
            >
              <button
                aria-haspopup="true"
                aria-label="Åpne meny"
                class="context-menu__button"
                data-id="context-menu-button"
                title="Åpne meny"
              >
                <ok-icon-stub
                  iconref="menu"
                  options="[object Object]"
                />
                <span
                  class="button-label"
                >
                  Valg
                </span>
              </button>
              <!--v-if-->
            </div>
          </nav>
        </header>
        <div
          class="graph__shadow"
          role="tabpanel"
        >
          <div
            aria-hidden="true"
            class="graph__container"
          >
            <div
              class="spinner"
            >
              <spinner-stub />
              <span
                class="spinner__text"
              >
                Henter data
              </span>
            </div>
            <!--v-if-->
            <svg
              aria-hidden="true"
              class="graph__svg loading"
            />
          </div>
          <div
            class="visually-hidden graph__tablecontainer"
          >
            <h3
              class="table-heading"
            />
            <table
              class=""
            >
              <thead />
              <tbody />
            </table>
            <!--v-if-->
          </div>
          <div
            class="vue3-resize-observer"
            tabindex="-1"
          >
            <object
              aria-hidden="true"
              data="about:blank"
              tabindex="-1"
              type="text/html"
            />
          </div>
          <!--v-if-->
        </div>
        <!--v-if-->
        <!--v-if-->
      </div>
    </section>
    <section
      class="card-container large"
    >
      <div
        class="card"
      >
        <!--v-if-->
        <header
          class="card__header"
        >
          <div
            class="card__headertext"
          >
            <h2
              class="card__title"
            >
              Gjennomsnitt- og medianalder
            </h2>
          </div>
          <nav
            class="card__nav"
          >
            <div
              class="tabs"
              role="tablist"
            >
              <div
                class="vue3-resize-observer"
                tabindex="-1"
              >
                <object
                  aria-hidden="true"
                  data="about:blank"
                  tabindex="-1"
                  type="text/html"
                />
              </div>
              
              <button
                aria-label="Status"
                class="active tabs__tab--hidden tabs__tab"
                data-id="tabButton-Gjennomsnitt- og medianalder-0"
                role="tab"
              >
                Status
              </button>
              
              <select
                aria-hidden=""
                class="select tabs__select"
              >
                
                <option
                  value="0"
                >
                  Status
                </option>
                
              </select>
            </div>
            <div
              class="card__toggle-menu"
            >
              <button
                aria-label="Graf"
                class="card__toggle-button card__toggle-button--active"
                title="Graf"
              >
                <ok-icon-stub
                  class="context-menu__dropdown-item-icon"
                  iconref="graph"
                  options="[object Object]"
                />
              </button>
              <button
                aria-label="Tabell"
                class="card__toggle-button"
                title="Tabell"
              >
                <ok-icon-stub
                  class="context-menu__dropdown-item-icon"
                  iconref="table"
                  options="[object Object]"
                />
              </button>
              <button
                aria-label="Kart"
                class="card__toggle-button"
                title="Kart"
              >
                <ok-icon-stub
                  class="context-menu__dropdown-item-icon"
                  iconref="location-pin"
                  options="[object Object]"
                />
              </button>
            </div>
            <div
              class="context-menu"
            >
              <button
                aria-haspopup="true"
                aria-label="Åpne meny"
                class="context-menu__button"
                data-id="context-menu-button"
                title="Åpne meny"
              >
                <ok-icon-stub
                  iconref="menu"
                  options="[object Object]"
                />
                <span
                  class="button-label"
                >
                  Valg
                </span>
              </button>
              <!--v-if-->
            </div>
          </nav>
        </header>
        <div
          class="graph__shadow"
          role="tabpanel"
        >
          <div
            aria-hidden="true"
            class="graph__container"
          >
            <div
              class="spinner"
            >
              <spinner-stub />
              <span
                class="spinner__text"
              >
                Henter data
              </span>
            </div>
            <!--v-if-->
            <svg
              aria-hidden="true"
              class="graph__svg loading"
            />
          </div>
          <div
            class="visually-hidden graph__tablecontainer"
          >
            <h3
              class="table-heading"
            />
            <table
              class=""
            >
              <thead />
              <tbody />
            </table>
            <!--v-if-->
          </div>
          <div
            class="vue3-resize-observer"
            tabindex="-1"
          >
            <object
              aria-hidden="true"
              data="about:blank"
              tabindex="-1"
              type="text/html"
            />
          </div>
          <!--v-if-->
        </div>
        <!--v-if-->
        <!--v-if-->
      </div>
    </section>
    
    <div
      class="related"
    >
      <h2
        class="section-heading"
      >
        Se også
      </h2>
      <div
        class="topics-grid"
      >
        
        <v-category-stub
          bgimage="/img/folkemengde"
          category="Befolkning"
          disabled="false"
          district="gamleoslo"
          id="befolkningsutvikling"
          topic="Befolkningsutvikling"
          txtcolor="rgb(245, 173, 165)"
        />
        <v-category-stub
          bgimage="/img/husholdninger"
          category="Befolkning"
          disabled="false"
          district="gamleoslo"
          id="husholdninger"
          topic="Husholdninger"
          txtcolor="rgb(245, 173, 165)"
        />
        <v-category-stub
          bgimage="/img/levekaar"
          category="Levekår"
          disabled="false"
          district="gamleoslo"
          id="levekaar"
          topic="Levekårs&shy;indikatorer"
          txtcolor="rgb(244, 206, 138)"
        />
        
      </div>
    </div>
  </div>
</div>
`;
